src/translate-adapter.ts

```diff
@@ -0,0 +1,455 @@
+	}
+
+	// i18n base file
+	if (options.base) {
+		i18nBases = options.base.map((p) => path.resolve(p));
+	} else {
+		const defaultPath = path.join(admin, "i18n", "en", "translations.json");
+		i18nBases = [
+			defaultPath,
+			path.join(admin, "src", "i18n", "en.json"),
+		].filter(existsSync);
+		if (i18nBases.length === 0) {
+			// if no path exists, we are most likely using words.js and
+			// expect the i18n file to be in the default path
+			i18nBases = [defaultPath];
+		}
+	}
+}
+
+/***************************** Command Handlers *******************************/
+
+async function handleTranslateCommand(): Promise<void> {
+	await translateIoPackage();
+	for (const i18nBase of i18nBases) {
+		await translateI18n(i18nBase);
+	}
+}
+
+async function handleToJsonCommand(): Promise<void> {
+	if (!existsSync(words)) {
+		return die(`Couldn't find words file ${words}`);
+	}
+
+	await adminWords2languages(words, i18nBases[0]);
+}
+
+async function handleToWordsCommand(): Promise<void> {
+	await adminLanguages2words(i18nBases[0]);
+}
+
+async function handleAllCommand(): Promise<void> {
+	await handleTranslateCommand();
+	await handleToWordsCommand();
+	await handleToJsonCommand();
+}
+
+/****************************** Implementation ********************************/
+
+async function translateIoPackage(): Promise<void> {
+	const content = await readJson(ioPackage);
+	if (content.common.news) {
+		console.log("Translate News");
+		for (const k in content.common.news) {
+			console.log(`News: ${k}`);
+			const nw = content.common.news[k];
+			await translateNotExisting(nw);
+		}
+	}
+	if (content.common.titleLang) {
+		console.log("Translate Title");
+		await translateNotExisting(
+			content.common.titleLang,
+			content.common.title,
+		);
+	}
+	if (content.common.desc) {
+		console.log("Translate Description");
+		await translateNotExisting(content.common.desc);
+	}
+	await writeJson(ioPackage, content, { spaces: 4, EOL });
+	console.log(`Successfully updated ${path.relative(".", ioPackage)}`);
+}
+
+async function translateNotExisting(
+	obj: Partial<Record<Language, string>>,
+	baseText?: string,
+): Promise<void> {
+	let text = obj.en;
+	if (!text) {
+		text = baseText;
+	}
+
+	if (text) {
+		for (const lang of getLanguages()) {
+			if (!obj[lang]) {
+				const time = new Date().getTime();
+				obj[lang] = await translateText(text, lang);
+				console.log(
+					gray(`en -> ${lang} ${new Date().getTime() - time} ms`),
+				);
+			}
+		}
+	}
+}
+
+async function translateI18n(baseFile: string): Promise<void> {
+	const filePattern = createFilePattern(baseFile);
+	const baseContent = await readJson(baseFile);
+	const missingLanguages = getLanguages();
+	const files = await findAllLanguageFiles(baseFile);
+	for (const file of files) {
+		const match = file.match(filePattern);
+		const lang = match![2] as Language; // language files always match
+		const langIndex = missingLanguages.indexOf(lang);
+		missingLanguages.splice(langIndex, 1);
```

Comment:
You could use a `Set` instead. Then this would just be `missingLanguages.delete(lang)`.